monai/apps/auto3dseg/bundle_gen.py

```diff
@@ -303,22 +307,78 @@ class BundleGen(AlgoGen):
-            # trigger the download process
-            zip_download_dir = TemporaryDirectory()
-            algo_compressed_file = os.path.join(zip_download_dir.name, "algo_templates.tar.gz")
-            download_and_extract(default_algo_zip if algos is None else algos, algo_compressed_file, algo_path)
-            zip_download_dir.cleanup()
-            sys.path.insert(0, os.path.join(algo_path, "algorithm_templates"))
-            algos = deepcopy(default_algos)
-            for name in algos:
-                algos[name]["template_path"] = os.path.join(
-                    algo_path, "algorithm_templates", algos[name]["template_path"]
-                )
+    def __init__(
+        self,
+        algo_path: str = ".",
+        algos: Optional[Union[Dict, List, str]] = None,
+        templates_path_or_url: Optional[str] = None,
+        data_stats_filename: Optional[str] = None,
+        data_src_cfg_name: Optional[str] = None,
+    ):
+
+        if algos is None or isinstance(algos, (list, tuple, str)):
+
+            if templates_path_or_url is None:
+                templates_path_or_url = default_algo_zip
+
+            algo_path = os.path.abspath(algo_path)
+            at_path = os.path.join(algo_path, "algorithm_templates")
+
+            if os.path.isdir(templates_path_or_url):
+                # copy, if a local folder
+                templates_path_or_url = os.path.abspath(templates_path_or_url)
+                if templates_path_or_url != at_path:
+                    if os.path.exists(at_path):
+                        shutil.rmtree(at_path)
+                    shutil.copytree(templates_path_or_url, at_path)
+
+                algos_all = {}
+                for name in os.listdir(at_path):
+                    if os.path.exists(os.path.join(templates_path_or_url, name, "scripts", "algo.py")):
+                        algos_all[name] = dict(
+                            _target_=f"{name}.scripts.algo.{name.capitalize()}Algo",
+                            template_path=os.path.join(at_path, name),
+                        )
+                if len(algos_all) == 0:
+                    raise ValueError(f"Unable to find any algos in {templates_path_or_url}")
+
+            elif urlparse(templates_path_or_url).scheme in ("http", "https"):
+                # trigger the download process if url
+                zip_download_dir = TemporaryDirectory()
+                algo_compressed_file = os.path.join(zip_download_dir.name, "algo_templates.tar.gz")
+
+                download_attempts = 3
+                for i in range(download_attempts):
+                    try:
+                        download_and_extract(templates_path_or_url, algo_compressed_file, algo_path)
+                    except Exception as e:
+                        msg = f"Download and extract of {templates_path_or_url} failed, attempt {i+1}/{download_attempts}."
+                        if i < download_attempts - 1:
+                            warnings.warn(msg)
```

Comment:
would be helpful to have `i` seconds of delay here to avoid too frequent requests. we ran into this issue https://github.com/Project-MONAI/MONAI/issues/5352

Reply:
ok, I've added the delays